demos/demo1-chat-heads/Demo.jsx

```diff
@@ -1,59 +1,222 @@
-  };
+    window.addEventListener("mousemove", this.handleMouseMove);
+    window.addEventListener("touchmove", this.handleTouchMove);
+    this.setState({ headerHeight: this.divRef.clientHeight });
+  },
+
+  shouldComponentUpdate(nextProps, nextState) {
+    if (this.state.x !== nextState.x) {
+      return true;
+    }
+    if (this.props.y !== nextProps.y) {
+      return true;
+    }
+    if (this.state.stiffness !== nextState.stiffness) {
+      return true;
+    }
+    if (this.state.damping !== nextState.damping) {
+      return true;
+    }
+    if (this.state.precision !== nextState.damping) {
+      return true;
+    }
+
+    return false;
+  },
 
-  handleMouseMove = ({pageX: x, pageY: y}) => {
-    this.setState({x, y});
-  };
+  handleMouseMove({ pageX: x, pageY: y }) {
+    const height = this.state.headerHeight;
+    const chatHeadRadiusPx = 20;
+    // chat heads avoid header area
+    if (y > height * 3.3 + chatHeadRadiusPx) {
+      this.setState({ x, y: y - height * 3.3 });
+    }
+  },
 
-  handleTouchMove = ({touches}) => {
+  handleTouchMove({ touches }) {
     this.handleMouseMove(touches[0]);
-  };
+  },
+
+  handleStiffnessChange(event) {
+    this.setState({ stiffness: event.target.value });
+  },
+
+  handlePrecisionChange(event) {
+    if (!isNaN(event.target.value)) {
+      this.setState({ precision: event.target.value });
+    }
+  },
 
-  getStyles = (prevStyles) => {
+  handleDampingChange(event) {
+    this.setState({ damping: event.target.value });
+  },
+
+  handleStiffnessParamChange(event) {
+    this.setState({ stiffnessBehaviour: event.target.value });
+  },
+
+  handleDampingParamChange(event) {
+    this.setState({ dampingBehaviour: event.target.value });
+  },
+
+  getStyles(prevStyles) {
     // `prevStyles` is the interpolated value of the last tick
+    let stiffness = this.state.stiffness;
+    let damping = this.state.damping;
+    let precision = this.state.precision;
+
     const endValue = prevStyles.map((_, i) => {
+      if (this.state.stiffnessBehaviour === "linear") {
+        stiffness = this.state.stiffness * i;
+      }
+      if (this.state.stiffnessBehaviour === "inverseLinear") {
+        stiffness = this.state.stiffness / i;
+      }
+      if (this.state.dampingBehaviour === "linear") {
+        damping = this.state.damping * i;
+      }
+      if (this.state.dampingBehaviour === "inverseLinear") {
+        damping = this.state.damping / i;
+      }
       return i === 0
         ? this.state
         : {
-          x: spring(prevStyles[i - 1].x, presets.gentle),
-          y: spring(prevStyles[i - 1].y, presets.gentle),
-        };
+            x: spring(prevStyles[i - 1].x, { stiffness, damping, precision }),
+            y: spring(prevStyles[i - 1].y, { stiffness, damping, precision })
+          };
     });
     return endValue;
-  };
+  },
 
   render() {
+    const selectOptions = options.map(({ value, description }) =>
```

Comment:
Make `selectOptions`, `stiffnessSelect`, `dampingSelect ` and `precisionInput` class methods, so they're not created as functions on each render (thus, cluttering the GC)?